### PR TITLE
DevTools: Parse named source AST in a worker

### LINKED ISSUE #21855: [DevTools] Parse named source AST in a worker
Hooks are the preferred way for writing stateful React components, but there are a few things about them that still lag behind the class component API: inspecting state in DevTools (since hooks aren’t “named”). Even now that this feature has been released, it remains disabled by default because the size of DEV bundles in larger apps (e.g. Facebook) makes parsing the AST very slow.

To alleviate this, we plan to provide an additional compilation tool that adds hook names to an extended source map during build time (#21782). However not every application will use this and so DevTools will still need to fall back to parsing source into an AST.

Currently this is done using the '@babel/parser' package on the main thread:
https://github.com/facebook/react/blob/c2c6ea1fdead7a33a33664675365b1575315276e/packages/react-devtools-extensions/src/parseHookNames.js#L536-L540

It would be nice to move this into a worker so that DevTools remains responsive even when parsing large sources.

---

One downside of using a worker for this is that serializing data to share between the worker and main thread (e.g. the source code string or the AST itself) could be expensive. This change will probably also require redesigning the boundaries a bit to reduce the number and size of things passed between boundaries. For example, the main thread could pass the URL of the source code to the worker, which could fetch and parse the file. Then the main thread could pass line/column information and the worker could return a hook name (or null). The goal would be to avoid having to serialize large AST structures.

### PR DESCRIPTION
**Pull request by tsirlucas**
## Summary

Resolves #21855

Ended up using [workerize](https://github.com/developit/workerize-loader) in order to setup the worker once it allows easy imports (for babel's parse function) and exports.

## Test Plan

- yarn build:chrome && yarn test:chrome --url=<URL_HERE>
- went to components tab on devtools
- checked out if the hooks names were correct
- changed the name of the hooks and tested again just in case

Also tested on Firefox, but for some reason needed to refresh the page after launching otherwise I would get a message telling me the website is not using React. Checked out on a previous commit and tested it and the same happens.

### CODE REVIEW

**Review by crypt096 — APPROVED** [review]

File: packages/react-devtools-extensions/package.json [hunk 1]
```diff
@@ -55,7 +55,8 @@
     "web-ext": "^3.0.0",
     "webpack": "^4.43.0",
     "webpack-cli": "^3.3.11",
-    "webpack-dev-server": "^3.10.3"
```
Comment (bvaughn) [inline packages/react-devtools-extensions/package.json:58]: Also don't forget to run `yarn install` in the root directory to update the `yarn.lock` file for this change.

```diff
diff --git a/yarn.lock b/yarn.lock
index c0bddfe82c..03daa69ff8 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -14877,6 +14877,13 @@ worker-loader@^3.0.2:
     loader-utils "^2.0.0"
     schema-utils "^2.7.0"
 
+workerize-loader@^1.3.0:
+  version "1.3.0"
+  resolved "https://registry.yarnpkg.com/workerize-loader/-/workerize-loader-1.3.0.tgz#4995cf2ff2b45dd6dc60e4411e63f5ae2c704d36"
+  integrity sha512-utWDc8K6embcICmRBUUkzanPgKBb8yM1OHfh6siZfiMsswE8wLCa9CWS+L7AARz0+Th4KH4ZySrqer/OJ9WuWw==
+  dependencies:
+    loader-utils "^2.0.0"
+
 wrap-ansi@^2.0.0:
   version "2.1.0"
   resolved "https://registry.yarnpkg.com/wrap-ansi/-/wrap-ansi-2.1.0.tgz#d8fc3d284dd05794fe84973caecdd1cf824fdd85"
```

File: packages/react-devtools-extensions/src/parseHookNames.js [hunk 2]
```diff
@@ -550,24 +551,26 @@ async function parseSourceAST(
...
-        console.log(
-          `parseSourceAST() Caching source metadata for "${originalSourceURL}"`,
-        );
-      }
-      originalURLToMetadataCache.set(originalSourceURL, {
-        originalSourceAST,
-        originalSourceCode,
-      });
+
+      promises.push(parsePromise);
```
[the remark this answers is not in the document]
Comment (tsirlucas) [reply packages/react-devtools-extensions/src/parseHookNames.js:570]: Didnt really tested it against large files so my bad. Gonna check and address the requested changes this evening.

[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/react-devtools-extensions/src/parseHookNames.js:570]: No worries at all! I really appreciate your help picking up this task 😄 Happy to talk more about it and help test the changes on my end.

**Comment by tsirlucas** [comment]
@bvaughn So, spent some time thinking about how to move only the necessary parts to the worker in order to keep the cache in the main thread and only send the url to the source files to the worker but noticed that in the end I would still need to send the `hooksList` object in order to call `findHookNames`. 

A solution would be also keeping I think keeping `findHookNames` in the main thread but thats also running some computation using the source code so it had to be kept in there and once this is true, I would have to send the `hooksList` to the worker anyways.

So I decided to just move the whole parser into the worker. It receives the `HooksTree` and returns the `HookNames`. I don't really have a huge project to test it out so can you give it a try?

Also fixed the other things you pointed out.

**Comment by bvaughn** [comment]
This feels _much_ more responsive after some initial testing! 😁 

I'm seeing a few components throw errors (failing to parse) that I don't think were failing before though, so I'm going to dig in tomorrow morning and see.

**Comment by tsirlucas** [comment]
> This feels _much_ more responsive after some initial testing! 😁 
> 
> 
> 
> I'm seeing a few components throw errors (failing to parse) that I don't think were failing before though, so I'm going to dig in tomorrow morning and see.

Perfect! If you can share some details on the components and erros so I can replicate things and debug would be great!

**Comment by bvaughn** [comment]
I think what's happening here is actually interesting. I think some components are timing out before the parsing is done, after your changes 😁  The cache has as built-in timeout, but before (since the work was all sync/blocking) the timer never had a chance to run. Now it does 😆  (I enabled the `__DEBUG__` logging and verified that this is what was happening.)

I think the solution is to just increase that timeout. It's really just there as a last resort in case something has gone really wrong. Let's increase it?
```diff
diff --git a/packages/react-devtools-shared/src/hookNamesCache.js b/packages/react-devtools-shared/src/hookNamesCache.js
index f385c7242..070dafeb1 100644
--- a/packages/react-devtools-shared/src/hookNamesCache.js
+++ b/packages/react-devtools-shared/src/hookNamesCache.js
@@ -18,7 +18,7 @@ import type {
 } from 'react-devtools-shared/src/types';
 import type {HookSource} from 'react-debug-tools/src/ReactDebugHooks';
 
-const TIMEOUT = 5000;
+const TIMEOUT = 30000;
 
 const Pending = 0;
 const Resolved = 1;
```

**Comment by tsirlucas** [comment]
> Now it does

@bvaughn done 😆

**Comment by bvaughn** [comment]
Chrome seems great now 👍🏼  but it looks like Firefox was broken by this change.

You can repro with:
```sh
 yarn build:firefox && yarn test:firefox --url=https://6tmfm.csb.app/
```

Enabling DEBUG mode might help track down where it fails too:
```diff
diff --git a/packages/react-devtools-extensions/src/parseHookNames/parseHookNames.js b/packages/react-devtools-extensions/src/parseHookNames/parseHookNames.js
index f68c37135..12ad6f76f 100644
--- a/packages/react-devtools-extensions/src/parseHookNames/parseHookNames.js
+++ b/packages/react-devtools-extensions/src/parseHookNames/parseHookNames.js
@@ -12,7 +12,7 @@ import LRU from 'lru-cache';
 import {SourceMapConsumer} from 'source-map';
 import {getHookName} from '../astUtils';
 import {areSourceMapsAppliedToErrors} from '../ErrorTester';
-import {__DEBUG__} from 'react-devtools-shared/src/constants';
+const __DEBUG__ = true;
 import {getHookSourceLocationKey} from 'react-devtools-shared/src/hookNamesCache';
 
 import type {
diff --git a/packages/react-devtools-shared/src/hookNamesCache.js b/packages/react-devtools-shared/src/hookNamesCache.js
index 070dafeb1..3e3d805cd 100644
--- a/packages/react-devtools-shared/src/hookNamesCache.js
+++ b/packages/react-devtools-shared/src/hookNamesCache.js
@@ -7,7 +7,7 @@
  * @flow
  */
 
-import {__DEBUG__} from 'react-devtools-shared/src/constants';
+const __DEBUG__ = true;
 
 import type {HooksTree} from 'react-debug-tools/src/ReactDebugHooks';
 import type {Thenable, Wakeable} from 'shared/ReactTypes';
```

**Comment by bvaughn** [comment]
As a worst-case fallback, we could parameterize the worker thing to only be used for Chrome and Edge (both of which work fine with this change).

File: packages/react-devtools-extensions/firefox/manifest.json [hunk 3]
```diff
@@ -32,7 +32,7 @@
 
   "devtools_page": "main.html",
 
-  "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'",
+  "content_security_policy": "script-src 'self' 'unsafe-eval' blob:; object-src 'self'",
```
Comment (bvaughn) [inline packages/react-devtools-extensions/firefox/manifest.json:35]: Oh right. Had to make the same change in #21897

**Review by bvaughn — APPROVED** [review]
Fantastic 🙇🏼 This feels much more responsive now. Thank you for the contribution! 👍🏼

**Pull request merged**